uqcsbot/channels.py

```diff
@@ -149,5 +197,62 @@ async def channel_manage_error(self, ctx: commands.context, error):
         else:
             logging.warning(error)
 
+    @commands.Cog.listener()
+    async def on_raw_reaction_add(self, payload):
+        """ Toggle adding/removing member from the corresponding channel. """
+        if payload.message_id == self.message_id:
+            guild = self.bot.get_guild(SERVER_ID)
+            member = guild.get_member(payload.user_id)
+
+            # Remove the reaction if not a bot.
+            channel = self.bot.get_channel(payload.channel_id)
+            msg = await channel.fetch_message(payload.message_id)
+            if member.bot:
+                return
+            await msg.remove_reaction(payload.emoji, member)
+                
+            # Covert to the correct format for custom emojis.
+            emoji_name = f"<:{payload.emoji.name}:{payload.emoji.id}>" if payload.emoji.id else payload.emoji.name
+
+            db_session = self.bot.create_db_session()
+            channel_query = db_session.query(Channel).filter(Channel.joinable == True, 
+                                                             Channel.emoji == emoji_name).one_or_none()
+            db_session.close()
+
+            if channel_query == None:
+                await member.send(f"Unable to find that channel.")
+                return
+
+            channel = self.bot.get_channel(channel_query.id)
+
+            if channel == None:
+                await member.send(f"Unable to find that channel.")
+                return
+
+            # Leave the channel if the user is currently a member.
+            if channel.permissions_for(member).is_superset(JOINED_PERMISSIONS):
+                await channel.set_permissions(member, read_messages=False, reason="UQCSbot removed.")
+                await member.send(f"You've left {channel.mention}")
+                return
+
+            # Otherwise, join the channel.
+            await channel.set_permissions(member, read_messages=True, reason="UQCSbot added.")
+            await member.send(f"You've joined {channel.mention}")
```

Comment:
Unsure if we necessarily need do DM the user every time they join and leave a channel. This was a point brought up a while ago as it may get annoying?

Reply:
I see your side of things there, however committee seems pretty set on the idea of having dms as confirmation. Since joining/leaving channels is a rare enough operation I think it is fine, but we can easily repeal this if reception is poor ;)

Reply:
If that is what committee decree, who am I to argue